File: go/stack/155_min_stack.go
package stack

// 设计一个支持 push ，pop ，top 操作，并能在常数时间内检索到最小元素的栈。
// Method 2: 不用辅助栈，额外的空间O(1)
// 思路: stack 中存放 当前push value 和 上次 lastMinVal 之间的差值 diff, 即是: diff = value - lastMinVal
// if diff < 0 => value 更小，他应该是当前这整个数组中最小的值，即minVal = value
// 否则，minValue 是数组中最小的值
// 这样Pop 的值如果是在栈顶，新的minVal = minVal - diff
type MinStack struct {
	data   []int
	minVal int
}

func Constructor() MinStack {
	// minVal 初始时长多少无所谓，后续push 元素的时候会判断初始情况
	return MinStack{
		data:   make([]int, 0),
		minVal: -1,
	}
}

func (m *MinStack) Push(val int) {
	if len(m.data) == 0 {
		// 初始时 存放原始值的差值是0
		m.data = append(m.data, 0)
		m.minVal = val
	} else {
		diff := val - m.minVal
		if diff < 0 {
			m.minVal = val
		}
		m.data = append(m.data, diff)
	}
}

func (m *MinStack) Pop() {
	if len(m.data) == 0 {
		return
	}

	diff := m.data[len(m.data)-1]
	// 栈顶是最小值元素，pop 之后需要更新 minVal
	if diff < 0 {
		m.minVal = m.minVal - diff
	}
	// 删除元素
	m.data = m.data[:len(m.data)-1]
}

func (m *MinStack) Top() int {
	if len(m.data) == 0 {
		return -1
	}
	diff := m.data[len(m.data)-1]
	// 栈顶即是最小值，value = minVal
	if diff < 0 {
		return m.minVal
	}
	return m.minVal + diff
}

func (m *MinStack) GetMin() int {
	return m.minVal
}

// 设计一个支持 push ，pop ，top 操作，并能在常数时间内检索到最小元素的栈。
// Method 1: 辅助栈，额外的空间O(n)
// type MinStack struct {
// 	stack    []int
// 	minStack []int
// }

// func Constructor() MinStack {
// 	stack := make([]int, 0)
// 	minStack := make([]int, 0)
// 	minStack = append(minStack, math.MaxInt64)
// 	return MinStack{
// 		stack:    stack,
// 		minStack: minStack,
// 	}
// }

// func (m *MinStack) Push(val int) {
// 	m.stack = append(m.stack, val)
// 	// FIX 防止index out of bound, 在Constructor中初始化一个最大值
// 	top := m.minStack[len(m.minStack)-1]
// 	// push 之后，minStack 都保存了stack 对应位置时的最小值
// 	m.minStack = append(m.minStack, min(top, val))
// }

// func (m *MinStack) Pop() {
// 	m.stack = m.stack[:len(m.stack)-1]
// 	m.minStack = m.minStack[:len(m.minStack)-1]
// }

// func (m *MinStack) Top() int {
// 	top := m.stack[len(m.stack)-1]
// 	return top
// }

// func (m *MinStack) GetMin() int {
// 	return m.minStack[len(m.minStack)-1]
// }

// func min(x, y int) int {
// 	if x < y {
// 		return x
// 	}

// 	return y
// }

File: rust/src/stack/_155_min_stack.rs
// O(1) 的空间复杂度, data 存储 diff = val - last_min_val
// 即: Top value = diff + last_min_val
// if diff < 0 => min_val = val
// i32 会越界，用i64
struct MinStack {
    data: Vec<i64>,
    min: i64,
}
/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {
    fn new() -> Self {
        Self {
            data: vec![],
            min: 0,
        }
    }

    fn push(&mut self, val: i32) {
        // 初始化时，min = 第一次的值，data 存储0
        if self.data.len() == 0 {
            self.data.push(0);
            self.min = val as i64
        } else {
            // 否则
            let diff = val as i64 - self.min;
            if diff < 0 {
                self.min = val as i64
            }
            self.data.push(diff)
        }
    }

    fn pop(&mut self) {
        if self.data.len() == 0 {
            return;
        }

        let diff = self.data.pop().unwrap();
        // 最后一个元素即是最小值，pop 后需要更新新的min_val
        if diff < 0 {
            self.min = self.min - diff
        }
    }

    fn top(&self) -> i32 {
        if self.data.len() == 0 {
            return -1;
        }

        let diff = self.data.last().unwrap();
        if *diff < 0 {
            return self.min as i32;
        } else {
            return (self.min + *diff) as i32;
        }
    }

    fn get_min(&self) -> i32 {
        self.min as i32
    }
}

// // O(n) 的空间复杂度, 需要一个额外的辅助栈，存储最小值
// struct MinStack {
//     stack: Vec<i32>,
//     min_stack: Vec<i32>,
// }

// /**
//  * `&self` means the method takes an immutable reference.
//  * If you need a mutable reference, change it to `&mut self` instead.
//  */
// impl MinStack {
//     fn new() -> Self {
//         let min_stack = vec![i32::MAX];
//         let stack = vec![];
//         MinStack { stack, min_stack }
//     }

//     fn push(&mut self, val: i32) {
//         self.stack.push(val);
//         let old_min = self.min_stack.last().unwrap();

//         use std::cmp::min;
//         self.min_stack.push(min(*old_min, val));
//     }

//     fn pop(&mut self) {
//         self.stack.pop();
//         self.min_stack.pop();
//     }

//     fn top(&self) -> i32 {
//         *self.stack.last().unwrap()
//     }

//     fn get_min(&self) -> i32 {
//         *self.min_stack.last().unwrap()
//     }
// }
